File: go/grimlock/version_manager.go
package grimlock

import (
	"fmt"
	"sync"

	"github.com/privyy/grimlock/types"
	v1 "github.com/privyy/grimlock/v1"
)

// VersionManager manages crypto version metadata and routing
type VersionManager struct {
	versions      map[string]*types.VersionMetadata
	latestVersion string
	compatibility types.CompatibilityMatrix
	mu            sync.RWMutex
}

var (
	instance *VersionManager
	once     sync.Once
)

// GetVersionManager returns the singleton version manager instance
func GetVersionManager() *VersionManager {
	once.Do(func() {
		instance = &VersionManager{
			versions:      make(map[string]*types.VersionMetadata),
			latestVersion: "v1",
			compatibility: make(types.CompatibilityMatrix),
		}
		instance.registerVersions()
	})
	return instance
}

// registerVersions registers all available versions
func (vm *VersionManager) registerVersions() {
	// Register v1
	vm.versions["v1"] = &types.VersionMetadata{
		Version: "v1",
		Algorithms: types.AlgorithmVersions{
			KeyExchange:   "X25519",
			KeyDerivation: "Argon2id-v13",
			Encryption:    "AES-256-GCM",
			HKDF:          "HKDF-SHA512",
			RecoveryKey:   "256-bit-random",
			Serialization: "Base64",
		},
		Constants:  v1.Constants,
		Deprecated: false,
	}

	// Set up compatibility matrix
	vm.compatibility["v1"] = map[string]bool{
		"v1": true,
		// v1 can read v1 data
	}

	// Future versions can be registered here
	// vm.versions["v2"] = &types.VersionMetadata{...}
}

// GetLatestVersion returns the latest version string
func (vm *VersionManager) GetLatestVersion() string {
	vm.mu.RLock()
	defer vm.mu.RUnlock()
	return vm.latestVersion
}

// GetVersion returns metadata for a specific version
func (vm *VersionManager) GetVersion(version string) (*types.VersionMetadata, error) {
	vm.mu.RLock()
	defer vm.mu.RUnlock()

	metadata, exists := vm.versions[version]
	if !exists {
		return nil, fmt.Errorf("version %s not found", version)
	}

	return metadata, nil
}

// IsCompatible checks if two versions are compatible
func (vm *VersionManager) IsCompatible(version1, version2 string) bool {
	vm.mu.RLock()
	defer vm.mu.RUnlock()

	compat, exists := vm.compatibility[version1]
	if !exists {
		return false
	}

	return compat[version2]
}

// ListVersions returns all registered version strings
func (vm *VersionManager) ListVersions() []string {
	vm.mu.RLock()
	defer vm.mu.RUnlock()

	versions := make([]string, 0, len(vm.versions))
	for v := range vm.versions {
		versions = append(versions, v)
	}
	return versions
}

// IsDeprecated checks if a version is deprecated
func (vm *VersionManager) IsDeprecated(version string) (bool, error) {
	metadata, err := vm.GetVersion(version)
	if err != nil {
		return false, err
	}
	return metadata.Deprecated, nil
}

// GetMigrationGuide returns the migration guide for a version
func (vm *VersionManager) GetMigrationGuide(version string) (string, error) {
	metadata, err := vm.GetVersion(version)
	if err != nil {
		return "", err
	}
	return metadata.MigrationGuide, nil
}

File: python/grimlock/version_manager.py
"""Version management utilities for Grimlock crypto module."""

from threading import Lock
from typing import Dict, Optional

from .types.version import (
    AlgorithmSpec,
    CompatibilityMatrix,
    CryptoConstants,
    VersionMetadata,
)
from .versions.v1.constants import CRYPTO_CONSTANTS_V1


class VersionManager:
    """Manages crypto version metadata and routing."""

    _instance: Optional["VersionManager"] = None
    _lock = Lock()

    def __init__(self):
        self._versions: Dict[str, VersionMetadata] = {}
        self._latest_version = "v1"
        self._compatibility: CompatibilityMatrix = {}
        self._register_versions()

    @classmethod
    def get_instance(cls) -> "VersionManager":
        """Get the singleton version manager instance."""
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = cls()
        return cls._instance

    def _register_versions(self) -> None:
        """Register all available versions."""
        # Register v1
        self._versions["v1"] = VersionMetadata(
            version="v1",
            algorithms=AlgorithmSpec(
                key_exchange="X25519",
                key_derivation="Argon2id-v13",
                encryption="AES-256-GCM",
                hkdf="HKDF-SHA512",
            ),
            constants=CRYPTO_CONSTANTS_V1,
            deprecated=False,
        )

        # Set up compatibility matrix
        self._compatibility["v1"] = {"v1": True}

    def get_latest_version(self) -> str:
        """Get the latest version string."""
        return self._latest_version

    def get_version(self, version: str) -> VersionMetadata:
        """Get metadata for a specific version.

        Args:
            version: Version string (e.g., "v1")

        Returns:
            VersionMetadata: Version metadata

        Raises:
            ValueError: If version not found
        """
        metadata = self._versions.get(version)
        if metadata is None:
            raise ValueError(f"version {version} not found")
        return metadata

    def is_compatible(self, version1: str, version2: str) -> bool:
        """Check if two versions are compatible.

        Args:
            version1: First version string
            version2: Second version string

        Returns:
            bool: True if versions are compatible
        """
        compat = self._compatibility.get(version1)
        if compat is None:
            return False
        return compat.get(version2, False)

    def list_versions(self) -> list[str]:
        """List all registered version strings.

        Returns:
            list[str]: List of version strings
        """
        return list(self._versions.keys())

    def is_deprecated(self, version: str) -> bool:
        """Check if a version is deprecated.

        Args:
            version: Version string

        Returns:
            bool: True if version is deprecated

        Raises:
            ValueError: If version not found
        """
        metadata = self.get_version(version)
        return metadata.deprecated

    def get_migration_guide(self, version: str) -> Optional[str]:
        """Get the migration guide for a version.

        Args:
            version: Version string

        Returns:
            Optional[str]: Migration guide or None

        Raises:
            ValueError: If version not found
        """
        metadata = self.get_version(version)
        return metadata.migration_guide


def get_version_manager() -> VersionManager:
    """Get the singleton version manager instance."""
    return VersionManager.get_instance()
